Add Drift Detection workshop (#162)
aws-samples/cfn101-workshop: workshop/content/intermediate/operations/drift-detection/_index.md

1. Navigate to the [CloudFormation Console](https://console.aws.amazon.com/cloudformation/). If necessary, choose the **Stacks** menu item to see your Stacks.
2. Choose your stack created in the earlier steps (for example, `drift-detection-challenge`).
3. From **Stack actions**, choose **Detect Drift**.
4. Drift detection takes a few moments to complete. Refresh the stack info page until the **Drift status** field shows `Drifted`.
5. From **Stack actions**, choose **View drift results**.
6. The drift status page displays, showing that `Instance1` has been modified.
7. Select `Instance1`, then choose **View drift details**.
8. The drift details show that the `UserData` property has been modified. The `UserData` property is stored using Base64 encoding, so the exact change you made is not obvious in the display.

{{% notice note %}}
You can use a tool to decode the Base64 text and see the shell script it represents. For example, on Linux you can use the `base64` command line tool to process it as follows.
Note that some implementations, for example macOS, use `-D` instead of `-d` as the option to the `base64` command.
{{% /notice %}}

> 

Reviewer: ```suggestion

For example, when you run the following command:

```

```shell

> $ echo "IyEvdXNyL2Jpbi9lbnYgYmFzaAplY2hvIEhlbGxvIFdvcmxkCg==" | base64 -d

Reviewer: ```suggestion
$ echo "IyEvdXNyL2Jpbi9lbnYgYmFzaAplY2hvIEhlbGxvIFdvcmxkCg==" | base64 -d
```
```

> #!/usr/bin/env bash

Reviewer: ```suggestion

should display the following output:

```text
#!/usr/bin/env bash
```

echo Hello World
```

{{% notice warning %}}
Be wary of using third-party websites to encode or decode your base64 text for UserData, especially if it contains or could contain sensitive information.
{{% /notice %}}

Your task now is to update the stack with the new state of the resource, without causing a further interruption.

{{%expand "Need a hint?" %}}
* You should delete the resource from the stack, then re-import it again with the `UserData` corrected. You do not need to convert the `UserData` to Base64.
* Refer to the [Resource Import](resource-importing.html) lab for more guidance.
{{% /expand %}}

{{%expand "Want to see the solution?" %}}

1. Update the `drift-detection-challenge.yaml` template to add a `DeletionPolicy` attribute with a value of `Retain` to the `Instance1` resource. Save the file.
2. Update the stack with the updated `drift-detection-challenge.yaml` template. This tells CloudFormation that when the resource is removed from the template, it should not delete it but just stop managing it.

> 3. Once the stack is updated, edit the template file again to remove the resource completely (you can comment it out, using the `#` character at the start of each line), and save the file.

Reviewer: ```suggestion
3. Once the stack update is complete, edit the template file again to remove the whole resource declaration (you can also choose to comment it out, using the `#` character at the start of each relevant line), and save the file.
```

```yaml
Resources:
  Instance1:
    DeletionPolicy: Retain
    Type: AWS::EC2::Instance
    Properties:
      ImageId: !Ref LatestAmiId
      InstanceType: t2.micro
      UserData: !Base64 |
        #!/usr/bin/env bash
        echo Hello Universe
```

6. Select the stack in the CloudFormation console, then from **Stack actions** choose **Import resources in to stack**.
7. Choose **Next**.
8. Upload the template file.
9. Enter the physical ID of the instance and choose **Next**.
10. In **Specify stack options**, choose **Next**.
11. Choose **Import resources**.
12. Once the stack operation is complete and the resource is imported, you can run drift detection on the stack to confirm the instance is now in sync with the stack template.

{{% /expand %}}

You can find the template for the solution in `code/solutions/drift-detection/drift-detection-workshop.yaml`.

Well done! You have now learned how to repair drift without impact by deleting and re-importing a resource.

### Cleanup

Follow steps shown next to clean up the resources you created in this workshop.

1. Navigate to the CloudFormation Console.

> 2. Choose the stack created in the first lab, for example `drift-detection-workshop`

Reviewer: ```suggestion
2. Choose the stack created in the first lab, for example `drift-detection-workshop`.
```

> 3. Choose **Delete**, then choose the **Delete stack**

Reviewer: ```suggestion
3. Choose **Delete**, then choose the **Delete stack**.
```

> 4. For the `drift-detection-challenge` stack, edit the template file to change the `DeletionPolicy` to `Delete`

Reviewer: ```suggestion
4. For the `drift-detection-challenge` stack, edit the template file to change the `DeletionPolicy` to `Delete`.
```
